File: _mcp/dedao-vault-mcp/dedao_vault_mcp/pages.py
import re
from dataclasses import dataclass, field
from pathlib import Path

import yaml

WIKILINK_RE = re.compile(r"\[\[([^\]]+)\]\]")


@dataclass
class Page:
    name: str
    path: str
    type: str | None
    aliases: list[str]
    tags: list[str]
    frontmatter: dict
    sections: dict
    body: str
    outlinks: list[str] = field(default_factory=list)


def parse_frontmatter(text: str) -> tuple[dict, str]:
    if text.startswith("---"):
        parts = text.split("---", 2)
        if len(parts) >= 3:
            fm = yaml.safe_load(parts[1]) or {}
            if not isinstance(fm, dict):
                fm = {}
            return fm, parts[2].lstrip("\n")
    return {}, text


def parse_sections(body: str) -> dict:
    sections: dict[str, str] = {}
    current = None
    buf: list[str] = []
    for line in body.splitlines():
        if line.startswith("## "):
            if current is not None:
                sections[current] = "\n".join(buf).strip()
            current = line[3:].strip()
            buf = []
        elif current is not None:
            buf.append(line)
    if current is not None:
        sections[current] = "\n".join(buf).strip()
    return sections


def parse_link_target(raw: str) -> str:
    s = raw.strip()
    if s.startswith("[[") and s.endswith("]]"):
        s = s[2:-2]
    s = s.split("|", 1)[0]
    s = s.split("#", 1)[0]
    return s.strip()


def extract_outlinks(body: str) -> list[str]:
    seen: list[str] = []
    for m in WIKILINK_RE.finditer(body):
        target = parse_link_target(m.group(1))
        if target and target not in seen:
            seen.append(target)
    return seen


def one_liner(sections: dict) -> str:
    for key in ("一句话定义", "简介"):
        text = (sections.get(key) or "").strip()
        if text:
            return text.splitlines()[0].strip()
    return ""


def parse_page(path: Path) -> Page:
    text = path.read_text(encoding="utf-8")
    fm, body = parse_frontmatter(text)
    return Page(
        name=path.stem,
        path=str(path),
        type=fm.get("type"),
        aliases=[a for a in (fm.get("aliases") or []) if isinstance(a, str)],
        tags=[t for t in (fm.get("tags") or []) if isinstance(t, str)],
        frontmatter=fm,
        sections=parse_sections(body),
        body=body,
        outlinks=extract_outlinks(body),
    )

File: _mcp/dedao-vault-mcp/src/pages.ts
import * as fs from "node:fs";
import * as path from "node:path";
import * as YAML from "yaml";

export interface Page {
  name: string;
  path: string;
  type: string | null;
  aliases: string[];
  tags: string[];
  frontmatter: Record<string, any>;
  sections: Record<string, string>;
  body: string;
  outlinks: string[];
}

// Use only with String.prototype.matchAll (never .exec()) — a shared /g regex carries lastIndex across exec calls.
export const WIKILINK_RE = /\[\[([^\]]+)\]\]/g;

function splitN(s: string, sep: string, n: number): string[] {
  const out: string[] = [];
  let rest = s;
  for (let i = 0; i < n - 1; i++) {
    const idx = rest.indexOf(sep);
    if (idx === -1) break;
    out.push(rest.slice(0, idx));
    rest = rest.slice(idx + sep.length);
  }
  out.push(rest);
  return out;
}

export function parseFrontmatter(text: string): [Record<string, any>, string] {
  if (text.startsWith("---\n")) {
    const parts = splitN(text, "---", 3);
    if (parts.length >= 3) {
      let fm: any = YAML.parse(parts[1]) ?? {};
      if (typeof fm !== "object" || Array.isArray(fm)) fm = {};
      return [fm, parts[2].replace(/^\n+/, "")];
    }
  }
  return [{}, text];
}

export function parseSections(body: string): Record<string, string> {
  const sections: Record<string, string> = {};
  let current: string | null = null;
  let buf: string[] = [];
  for (const line of body.split("\n")) {
    if (line.startsWith("## ")) {
      if (current !== null) sections[current] = buf.join("\n").trim();
      current = line.slice(3).trim();
      buf = [];
    } else if (current !== null) {
      buf.push(line);
    }
  }
  if (current !== null) sections[current] = buf.join("\n").trim();
  return sections;
}

export function parseLinkTarget(raw: string): string {
  let s = raw.trim();
  if (s.startsWith("[[") && s.endsWith("]]")) s = s.slice(2, -2);
  s = s.split("|")[0];
  s = s.split("#")[0];
  return s.trim();
}

export function extractOutlinks(body: string): string[] {
  const seen: string[] = [];
  for (const m of body.matchAll(WIKILINK_RE)) {
    const target = parseLinkTarget(m[1]);
    if (target && !seen.includes(target)) seen.push(target);
  }
  return seen;
}

export function oneLiner(sections: Record<string, string>): string {
  for (const key of ["一句话定义", "简介"]) {
    const text = (sections[key] ?? "").trim();
    if (text) return text.split("\n")[0].trim();
  }
  return "";
}

export function parsePage(filePath: string): Page {
  const text = fs.readFileSync(filePath, "utf-8");
  const [fm, body] = parseFrontmatter(text);
  const aliases = Array.isArray(fm.aliases) ? fm.aliases.filter((a: any) => typeof a === "string") : [];
  const tags = Array.isArray(fm.tags) ? fm.tags.filter((t: any) => typeof t === "string") : [];
  return {
    name: path.basename(filePath, ".md"),
    path: filePath,
    type: typeof fm.type === "string" ? fm.type : null,
    aliases,
    tags,
    frontmatter: fm,
    sections: parseSections(body),
    body,
    outlinks: extractOutlinks(body),
  };
}
